### PR TITLE
SSD MVP

### PR DESCRIPTION
**Pull request by wmww**
Not based on other PRs, but won't take effect until combined with #1119. After that decorations should appear on all X11 windows. Adds a working title bar with resize edges and minimize/maximize/close buttons.

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
As well as the compiler errors, resizing needs to be constrained so that the client window has a sane size.

File: src/server/shell/decoration/basic_decoration.cpp [hunk 1]
```diff
@@ -158,6 +158,8 @@ msd::BasicDecoration::BasicDecoration(
     std::shared_ptr<ms::Surface> const& window_surface)
     : threadsafe_self{std::make_shared<ThreadsafeAccess<BasicDecoration>>(this, executor)},
       static_geometry{std::make_unique<StaticGeometry>(default_geometry)},
+      window_state{nullptr},
+      input_state{nullptr},
```
Comment (AlanGriffiths) [inline src/server/shell/decoration/basic_decoration.cpp]: Why? The default initialization does the right thing.

Comment (wmww) [reply src/server/shell/decoration/basic_decoration.cpp]: I just want to make it explicit that null initialization isn't a mistake.

Comment (AlanGriffiths) [reply src/server/shell/decoration/basic_decoration.cpp]: :man_shrugging:

**Comment by AlanGriffiths** [comment]
I think what's missing is mapping from the size constraints set for the client_size to constraints for the window_size seen by Window Management. That hasn't been a distinction before, but becomes one with decorations.

From a quick examination that needs a chunk of infrastructure that could be landed separately (and first).

**Comment by wmww** [comment]
Size constraints don't work at all with XWayland surfaces, and I don't know why. Doesn't matter for now. Crashes seemed to be caused by logic bugs in the decoration, which have now been fixed.

**Comment by AlanGriffiths** [comment]
The latest changes do indeed prevent a crash, but they don't fix the logic that leads to the crash.

Take a simple X11 application (I'm using xeyes) and drag the bottom margin up: you can drag it into the titlebar. The resize ought to be constrained before that happens (try it with a GTK or Qt app.)

The same is true of dragging the right margin into the left etc. (This also leads to weird visuals: the horizontal margins are not as wide as the vertical ones.)

**Comment by wmww** [comment]
> they don't fix the logic that leads to the crash

Lead to the crash (past tense). The crash was caused by a decoration drawing bug, which is fixed.

As I said, size constraints seem to have no effect on XWayland windows (with or without decorations). This just wasn't apparent before because there was no way to resize them. This is an issue for another day.

**Comment by wmww** [comment]
Update: it appears max/min size are ignored for all surfaces except when they are initially set. Wayland surfaces appear to have a min size only because the client refuses to draw smaller buffers, and we don't clip.

**Comment by wmww** [comment]
Update: I fixed max/min size constraints on Friday and will PR that once I write tests on Monday or Tuesday

**Pull request merged**